From 3e0f6cb81a320532312fbe5233bf21648b043ecb Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Fri, 1 Dec 2017 14:48:09 +0000
Subject: [PATCH] archive: Take care to remove old pip installs when installing
 new

This solves solus-project/solus-sc#165 by nuking the old pip installed
tree when installing the new egg-info, and is highly specific to the
Python2 setuptools. This workaround will suffice until sol comes along.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 pisi/archive.py | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/pisi/archive.py b/pisi/archive.py
index 196ead1..831369b 100644
--- a/pisi/archive.py
+++ b/pisi/archive.py
@@ -268,6 +268,20 @@ class ArchiveTar(ArchiveBase):
         super(ArchiveTar, self).unpack(target_dir, clean_dir)
         self.unpack_dir(target_dir)
 
+    def maybe_nuke_pip(self, info):
+        if not info.name.endswith(".egg-info"):
+            return
+        if not "site-packages" in info.name:
+            return
+        if not "/python" in info.name:
+            return
+        if not info.isreg():
+            return
+        if not os.path.isdir(info.name):
+            return
+        print("Overwriting stale pip install: /{}".format(info.name))
+        shutil.rmtree(info.name)
+
     def unpack_dir(self, target_dir, callback=None):
         rmode = ""
         self.tar = None
@@ -301,6 +315,12 @@ class ArchiveTar(ArchiveBase):
             if callback:
                 callback(tarinfo, extracted=False)
 
+            try:
+                self.maybe_nuke_pip(tarinfo)
+            except Exception as e:
+                print("Failed to remove stale pip install: {}".format(e))
+                raise e
+
             if tarinfo.issym() and \
                     os.path.isdir(tarinfo.name) and \
                     not os.path.islink(tarinfo.name):
-- 
2.15.1

